src/main/java/com/hierynomus/smbj/transport/tcp/AsyncDirectTcpTransport.java

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright (C)2016 - SMBJ Contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.hierynomus.smbj.transport.tcp;
+
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.nio.channels.AsynchronousChannelGroup;
+import java.nio.channels.AsynchronousSocketChannel;
+import java.nio.channels.CompletionHandler;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.hierynomus.protocol.Packet;
+import com.hierynomus.protocol.commons.buffer.Buffer;
+import com.hierynomus.protocol.commons.buffer.Buffer.BufferException;
+import com.hierynomus.smbj.lock.CrossThreadLock;
+import com.hierynomus.smbj.transport.PacketHandlers;
+import com.hierynomus.smbj.transport.TransportException;
+import com.hierynomus.smbj.transport.TransportLayer;
+
+/**
+ * A transport layer over Direct TCP/IP that uses asynchronous I/O.
+ */
+public class AsyncDirectTcpTransport<P extends Packet<P, ?>> implements TransportLayer<P> {
+    public static final int DEFAULT_CONNECT_TIMEOUT = 5000;
+
+    private final Logger logger = LoggerFactory.getLogger(this.getClass());
+    private final PacketHandlers<P> handlers;
+    private final AsynchronousSocketChannel socketChannel;
+    private final AsyncPacketReader<P> packetReader;
+    private int soTimeout = 0;
+
+    // AsynchronousSocketChannel doesn't support concurrent writes
```

Comment:
@pepijnve Many elements in the existing transport and SMB layer design are already asynchronous. The new suggested mode is really about shifting this asynchronous design from a thread-per-connection model to a shared thread model.

On the transport layer, the "read" side of the existing transport is already implemented asynchronously, using a dedicated thread per connection. On the SMB layer, every operation returns a Future object, which is really an async design.

The only difference in functionality with the new transport, as you point out, is that now writes return early. I think the only impact of this difference is what happens when a write fails. Since there is already a facility in place for reporting asynchronous errors, currently named PacketReader, the async implementation simply reuses it also for reporting async write errors.
This raises a valid concern about naming and interface responsibilities, since write errors get reported to a PacketReader. I think maybe it's time to refactor PacketReader into two separate interfaces - PacketReader and a TransportErrorHandler. The TransportErrorHandler part would be used for reporting any asynchronous transport error, for both reading and writing. Does that make sense?

Is there any other use case where you expect things to break by by changing the write model?